core/templates/dev/head/pages/exploration-editor-page/services/exploration-rights.service.spec.ts

```diff
@@ -27,73 +27,275 @@ import { UpgradedServices } from 'services/UpgradedServices';
+
+      $httpBackend.expectPUT('/createhandler/rights/12345')
+        .respond(500);
+      ers.setViewability(true).then(successHandler, failHandler);
+      $httpBackend.flush();
+
+      expect(ers.viewableIfPrivate()).toBe(false);
+      expect(clearWarningsSpy).not.toHaveBeenCalled();
+      expect(successHandler).not.toHaveBeenCalled();
+      expect(failHandler).toHaveBeenCalled();
     });
 
-    it('reports the correct derived statuses', function() {
-      ers.init(['abc'], [], [], [], 'private', 'e1234', true);
-      expect(ers.isPrivate()).toBe(true);
-      expect(ers.isPublic()).toBe(false);
+  it('should save a new member', function() {
+    var sampleDataResultsCopy = angular.copy(sampleDataResults);
+    sampleDataResultsCopy.rights.viewer_names.push('newUser');
+
+    $httpBackend.expectPUT('/createhandler/rights/12345').respond(
+      200, sampleDataResultsCopy);
+    ers.saveRoleChanges('newUser', 'viewer').then(function() {
+      expect(ers.viewerNames).toEqual(
+        sampleDataResultsCopy.rights.viewer_names);
+    });
+    $httpBackend.flush();
+  });
 
-      ers.init(['abc'], [], [], [], 'public', 'e1234', true);
-      expect(ers.isPrivate()).toBe(false);
+  it('should reject handler when saving a new member fails', function() {
+    var successHandler = jasmine.createSpy('success');
+    var failHandler = jasmine.createSpy('fail');
+
+    $httpBackend.expectPUT('/createhandler/rights/12345')
+      .respond(500);
+    ers.saveRoleChanges('newUser', 'viewer').then(successHandler, failHandler);
+    $httpBackend.flush();
+
+    expect(clearWarningsSpy).not.toHaveBeenCalled();
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalled();
+  });
+
+  it('should make exploration rights public', function() {
+    var sampleDataResultsCopy = angular.copy(sampleDataResults);
+    sampleDataResultsCopy.rights.status = 'public';
+
+    $httpBackend.expectPUT('/createhandler/status/12345').respond(
+      200, sampleDataResultsCopy);
+    ers.publish(true).then(function() {
       expect(ers.isPublic()).toBe(true);
     });
+    $httpBackend.flush();
   });
+
+  it('should reject handler when making exploration rights public fails',
+    function() {
+      var successHandler = jasmine.createSpy('success');
+      var failHandler = jasmine.createSpy('fail');
+
+      $httpBackend.expectPUT('/createhandler/status/12345')
+        .respond(500);
+      ers.publish(true).then(successHandler, failHandler);
+      $httpBackend.flush();
+
+      expect(clearWarningsSpy).not.toHaveBeenCalled();
+      expect(successHandler).not.toHaveBeenCalled();
+      expect(failHandler).toHaveBeenCalled();
+    });
+
+  it('should save moderator change to backend', function() {
+    $httpBackend.expectPUT('/createhandler/moderatorrights/12345').respond(
+      200, sampleDataResults);
+    ers.saveModeratorChangeToBackend().then(function(data) {
+      expect(clearWarningsSpy).toHaveBeenCalled();
```

Comment:
You aren't doing anything with the variable "data"?

Reply:
No, I've removed it.